File: windows/src/GoalMaker.Core/Updates/UpdateService.cs
using GoalMaker.Core.Versioning;

namespace GoalMaker.Core.Updates;

/// <summary>
/// Release discovery → signature and content check → version policy → verified download → installer
/// launch, each behind its own seam (CodePrint updater rule). Never touches user data.
/// </summary>
public sealed class UpdateService(
    string installedVersion,
    ReleasePlatform platform,
    bool channelConfigured,
    IReleaseChannel channel,
    ReleaseVerifier verifier,
    IUpdateInstaller installer)
{
    public async Task<UpdateCheckResult> CheckAsync(CancellationToken cancellationToken)
    {
        if (!channelConfigured)
        {
            return new UpdateCheckResult.NotConfigured();
        }

        if (SemanticVersion.Parse(installedVersion) is not { IsDevelopmentBuild: false })
        {
            return new UpdateCheckResult.DevelopmentBuild();
        }

        ChannelSnapshot snapshot;
        try
        {
            snapshot = await channel.FetchLatestAsync(cancellationToken).ConfigureAwait(false);
        }
        catch (Exception error) when (error is not OperationCanceledException)
        {
            return new UpdateCheckResult.Failed(error.Message);
        }

        if (verifier.Check(snapshot) is not ManifestCheck.Valid valid)
        {
            return new UpdateCheckResult.Untrusted();
        }

        var manifest = valid.Manifest;
        var artifact = manifest.ArtifactFor(platform);
        return artifact is not null && UpdatePolicy.ShouldOffer(installedVersion, manifest.Version.ToString())
            ? new UpdateCheckResult.Available(manifest, artifact)
            : new UpdateCheckResult.UpToDate(manifest.Version.ToString());
    }

    public async Task<InstallResult> InstallAsync(
        UpdateCheckResult.Available update,
        IProgress<double>? progress,
        CancellationToken cancellationToken)
    {
        var expected = update.Artifact;
        var bytesProgress = progress is null
            ? null
            : new Progress<long>(read => progress.Report(Math.Clamp((double)read / expected.Size, 0, 1)));
        DownloadedArtifact downloaded;
        try
        {
            downloaded = await channel.DownloadAsync(expected.Path, bytesProgress, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception error) when (error is not OperationCanceledException)
        {
            return new InstallResult.Failed(error.Message);
        }

        if (downloaded.Size != expected.Size
            || !string.Equals(downloaded.Sha256, expected.Sha256, StringComparison.OrdinalIgnoreCase))
        {
            return new InstallResult.DownloadCorrupted();
        }

        installer.Launch(downloaded.LocalPath);
        return new InstallResult.InstallerStarted();
    }
}

File: android/app/src/main/kotlin/com/goalmaker/app/application/update/UpdateService.kt
package com.goalmaker.app.application.update

import com.goalmaker.app.domain.update.ManifestCheck
import com.goalmaker.app.domain.update.ReleasePlatform
import com.goalmaker.app.domain.update.UpdatePolicy
import com.goalmaker.app.domain.version.SemanticVersion
import kotlinx.coroutines.CancellationException

/**
 * Release discovery → signature and content check → version policy → verified download → installer
 * launch, each behind its own seam (CodePrint updater rule). Never touches user data.
 */
class UpdateService(
    private val installedVersion: String,
    private val platform: ReleasePlatform,
    private val channelConfigured: Boolean,
    private val channel: ReleaseChannel,
    private val verifier: ReleaseVerifier,
    private val installer: UpdateInstaller,
) {
    suspend fun check(): UpdateCheckResult {
        if (!channelConfigured) return UpdateCheckResult.NotConfigured
        if (SemanticVersion.parse(installedVersion)?.isDevelopmentBuild != false) {
            return UpdateCheckResult.DevelopmentBuild
        }
        val snapshot = try {
            channel.fetchLatest()
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (error: Exception) {
            return UpdateCheckResult.Failed(error.message ?: error::class.simpleName.orEmpty())
        }
        val manifest = when (val check = verifier.check(snapshot)) {
            is ManifestCheck.Valid -> check.manifest
            ManifestCheck.BadSignature, is ManifestCheck.BadManifest -> return UpdateCheckResult.Untrusted
        }
        val artifact = manifest.artifactFor(platform)
        return if (artifact != null && UpdatePolicy.shouldOffer(installedVersion, manifest.version.toString())) {
            UpdateCheckResult.Available(manifest, artifact)
        } else {
            UpdateCheckResult.UpToDate(manifest.version.toString())
        }
    }

    suspend fun install(update: UpdateCheckResult.Available, onProgress: (Float) -> Unit = {}): InstallResult {
        val expected = update.artifact
        val downloaded = try {
            channel.download(expected.path) { read -> onProgress((read.toFloat() / expected.size).coerceIn(0f, 1f)) }
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (error: Exception) {
            return InstallResult.Failed(error.message ?: error::class.simpleName.orEmpty())
        }
        if (downloaded.size != expected.size || !downloaded.sha256.equals(expected.sha256, ignoreCase = true)) {
            return InstallResult.DownloadCorrupted
        }
        installer.launch(downloaded.localPath)
        return InstallResult.InstallerOpened
    }
}
